### ISSUE TITLE
fs2.Sink/Channel replacement

### ISSUE THREAD

**Issue by pchlupacek** [functional-streams-for-scala/fs2#501]
Rethink and properly describe the Channel/Sink replacement in fs2

**Comment by pchlupacek** [comment]
I know we have been discussing this on few places, I just opened this issue so we may correctly track it. Hence channel and sink from fs2._ is widely used in almost any program, I think we need to clearly describe how this will be done in fs2._ 

I specifically curious how we would do state-dependent sinks and channel alternatives, that first acquired resource and then produced stream of functions. 

Also would like to see, if and how we can combine channel/sink replacements. I just want to make sure concept is well understood before implementing them on async primitives.

**Comment by pchiusano** [comment]
Yeah agree this is still open. I'm going to do a writeup of some possible
designs early this week and will post here and we can discuss.

I think we can hold off on implementing any conversion functions between
the async types and channels until we are solid on the channel
representation.

**Comment by pchlupacek** [comment]
Sure agree. I'll implement them once we will know more on this topic.

**Comment by pchiusano** [comment]
_Note:_ not bothering to discuss sinks here, which I think will always just be `type Sink[F[_],-A] = Channel[F,A,Unit]` or `Channel[F,A,Nothing]`, regardless of representation of `Channel`.

### Option 1

```Scala
type Channel[F[_],-A,+B] = Stream[F,A] => Stream[F,B]
```

_The good:_ It's very flexible in one sense - every operation available on streams is available when building up a channel. Channels can be a simple lift of some `A => F[B]`, or might pick an `A => F[B]` based on what `A` is received. The channel can insert or delete `A` values, transform them in any way, and stateful channels are possible, for things like rotating loggers, etc.

_The bad:_ It's very opaque and is missing a story for how to _interleave_ sending values through multiple channels. For example, how do we implement `observe`:

```Scala
def observe[F[_],A,B](s: Stream[F,A], sink: Sink[F,A]): Stream[F,A]
```

Likewise for more complicated scenarios, like sending the same value through two channels and combining the results in some fashion.

An observation here is that it is actually a general limitation of the library, that it has no story for how to _replicate/fork streams_ for consumption in multiple places (short of using mutable state like queues and so forth). So addressing this problem in a general way might be good---it lets us use this very flexible and simple representation of channels, with interleaving solved by a more general capability (which we still have to design).

Let's come back to this one.

### Option 2 (rejected)

Here's one I was playing with based on some discussion with @mpilquist on scodec-stream:

```Scala
trait Channel[F[_],-A,+B] { self =>
  type S
  def s0: S
  def step: S => Pull[F,B,Chunk[A] => S]
}

def through[F[_],A,B](s: Stream[F,A])(c: Channel[F,A,B]): Stream[F,B] = {
  def go[S](h: Handle[F,A], s0: S, step: S => Pull[F,B,Chunk[A] => S]): Pull[F,B,Nothing] =
    step(s0) flatMap { recv =>
      h.receive { case a #: h => go(h, recv(a), step) }
    }
  s pull { h => go(h, c.s0, c.step) }
}
```

It's kind of interesting - rather than being the entire stream transformation, it's supposed to just be one step of the transformation, which returns a new state. There are several variations of this basic idea.

The good part is that many operations like `observe` and `lift` are easy to implement:

```Scala
def observe[F[_],A](s: Stream[F,A])(c: Sink[F,A]): Stream[F,A] = {
  def go[S](h: Handle[F,A], s0: S, step: S => Pull[F,Nothing,Chunk[A] => S]): Pull[F,A,Nothing] =
    step(s0) flatMap { recv =>
      h.receive { case a #: h => Pull.output(a) >> go(h, recv(a), step) }
    }
  s pull { h => go(h, c.s0, c.step) }
}

def lift[F[_],A,B](f: A => F[B]): Channel[F,A,B] = new Channel[F,A,B] {
  type S = Chunk[A]
  val s0 = Chunk.empty
  def step = _.uncons match {
    case None => Pull.pure(identity)
    case Some((a,s)) => Pull.eval(f(a)).flatMap(Pull.output1) >> step(s)
  }
}
```

* Channels can evaluate effects, but cannot acquire resources... unless we make resource acquisition part of the public `Pull` api. Instead you'd have to work with a `Stream[F,Channel[F,A,B]]` where the outer stream does the resource acquisition.
* No way to `map` or `pipe` a `Channel`. Channels instead have to be applied to a stream to produce another stream, which another channel may then act on. Could be fixed by adding a function to operate on the output parameter of a `Pull`, but that adds complexity.
* Representation feels kind of ad hoc. Lots of things may be expressed, but it's done in a very different way than the rest of the library.

### Option 3, the status quo ###

We could keep the current representation, or switch to the slight variation:

```Scala
type Channel[F[_],-A,+B] = Stream[F, A => Stream[F,B]]
```

Besides needing a bunch of ad hoc combinators for working with channels, the problem with this representation is that there's no way for _evaluation of effects to determine how future inputs to the channel are processed_. This ends up being very limiting - a rotating logfile might check the file size after each write, but with this signature there's no way for this to influence how the next element gets processed... short of using mutable state.

### Option 1a, forking and replication combinators

I like option 1. It's the most expressive, the simplest to implement (it's done already, yay!) and it has just one downside---no support for interleaving. But this downside is a problem with the library more generally. So let's try to solve that. I have some thoughts on it that I'll try to write up very soon...

**Comment by pchiusano** [comment]
I posted a design [here in this gist](https://gist.github.com/pchiusano/2f60860447f2fbfa5979). It compiles. Naming and organization TBD, that is just a strawman.

Here's the key function:

```Scala
  /** 
   * Pass elements of `s` through both `f` and `g`, then combine the two resulting streams.
   * Implemented by enqueueing elements as they are seen by `f` onto a `Queue` used by the `g` branch.
   * USE EXTREME CARE WHEN USING THIS FUNCTION. Deadlocks are possible if `combine` pulls from the `g` 
   * branch synchronously before the queue has been populated by the `f` branch.
   * 
   * The `combine` function receives an `F[Long]` effect which evaluates to the current size of the
   * `g`-branch's queue.
   * 
   * When possible, use one of the safe combinators like `[[observe]]`, which are built using this function,
   * in preference to using this function directly.
   */
def diamond[F[_],A,B,C,D](s: Stream[F,A])
    (f: Stream[F,A] => Stream[F,B])
    (qs: QueueStrategy[Chunk[A]], g: Stream[F,A] => Stream[F,C])
    (combine: (Stream[F,B], F[Long], Stream[F,C]) => Stream[F,D])(implicit F: Async[F]): Stream[F,D]
```

This can be used to implement `observe` and pretty much whatever else. It's very flexible... and also potentially dangerous, as the comment indicates. However, we can provide safe combinators like `observe` and whatever else, and just leave the general function there for people who need to do something weird... at which point they should know what they are doing.

Some comments:

* This means that `observe` and some other combinators that formerly needed to know nothing about `F` now require an `AsyncExt[F]`. I don't think this is super problematic. If you were working with a non async `F` and using `observe`, chances are good you can phrase what you are doing another way.
* We don't pass in a `Queue` directly, just a 'recipe' for creating one. This way `diamond` handles the lifecycle of the queue, which is specific to this diamond being created. If you passed in a queue you were referencing elsewhere, that would result in bizarre spooky action at a distance, so we disallow that.
* Notice the `f` branch doesn't have a queueing strategy. That's because the `f` branch drives everything - it is the sole branch that ever reads directly from the original input stream. The `g` branch cannot run unless the `f` branch has enqueued some values for it to work with. More on that choice below.
* As a consequence of the `f` branch driving everything, `g` will never receive any uncaught exceptions from the original stream. These will just manifest as end-of-stream in the input to `g`. Hence, any errors will come to the `combine` function via the result of the `f` branch.

Overall I'm reasonable happy with this basic design, but am interested to hear what others think.

I can prepare a PR with this and various combinators implemented using `diamond` once we merge #462.

Some things I considered and rejected:

* A more symmetric `diamond` combinator, where both branches have a queue. When awaiting from a branch, if the queue is empty, it awaits from the original stream and enqueues the chunk it gets back in the other branch's queue. This starts to fall apart / become very scary when resources are involved. Who owns the resource? When can it be freed?
  * I can actually imagine doing some form of reference counting, to allow a resource to be shared by two branches, and freed only when both are done with it. But I rejected this as adding too much complexity to the interpreter.
  * It is possible there are nice answers to the question of how to have a symmetric diamond operation with resource acquisition happening in the input stream, but it seems better to leave that for 'future work'.
* I pondered whether it would be possible to generate nice error messages if deadlock occurs due to a bad `combine` function. The implementation I gave in the gist will just hang waiting on a `dequeue` that will never finish. But failing fast with a nice error is not possible as far as I can tell. The problem is that `combine` may be running one or both branches asynchronously. If the queue is empty, that isn't always a deadlock, since the other branch may be stepping asynchronously to fill it, or the branch which reads from the queue may be running asynchronously, which permits the other branch to run synchronously (or asynchronously) at a later time to fill the queue.

**Comment by pchiusano** [comment]
Something I just thought of is that this design means that `async` and `core` need to be in the same module... unless we wanted to move combinators like `observe` out of `core`. Not sure what I think about that.

**Comment by pchlupacek** [comment]
@pchiusano I'll look that over weekend. I don't have super good feeling about diamond though, can't at first imagine any users will use it. I need to grasp your writeup first though. thanks a lot!

**Comment by pchiusano** [comment]
Cool. Yeah diamond is meant to be a low level combinator just used to
implement other stuff like observe.

**Comment by pchlupacek** [comment]
@pchiusano I thought about this and I think good place to start is to see how we would write certain scenarios with new design, so we can prove that this will work before we will settle this channel/ sink and implement. I tried to collect use cases of sink and channel from our code and various combinations. If you will give me a hint how you would like them to appear, then I can do some small chapter to new design explaining the effect full channels and sinks.  

In general I don't have super good feeling in diamond operator. Not just its signature is too complex, but also that it involves asynchronous operation. This in fact means the effects will happen on different thread and we have used a lot  in our code the fact that process may either fail or succeed through sinks. Not sure if that will hold anymore with diamond.  

(TSink and TSource are just aliases for Sink[Task,A], Process[Task,A]...)

1. simple sink with resource acquire and release
```scala
val resourceSink: TSink[Array[Byte]] =  resource(acquire)(step)(release) 

val src:TSource[Array[Byte]] = ???

src.to(resourceSink)
```

2. combining channels 
```scala
// we combine two sinks here, where we want to have effect B to be scheduled after effect A 
val channel1: TChannel[A, B] = ???
val sink2: TSink[B] = ???
val sink3: TSink[C] =  (channel1 zip sink2).map { case (fa, fb) => (c:C) => fa(c.getA) flatMap { b =>  fb(b) }}
```

3. observing
```scala 
val sinkA:TSink[A] = ???
val channel:TSink[A,B] = ???
val src:TSource[A] = ???

src.observe(sinkA)
.observeThrough(channel)
:TSource[(A,B)]

```

**Comment by pchiusano** [comment]
These are good scenarios.

I understand the not wanting to have the sink / channel run in a separate thread or involve. I guess I have sort of justified it to myself by reasoning that you can do things synchronously in lots of other ways. For instance, if you just have an effectful function which is backed by some resource, just use bracket for that directly. If you just have an effectful function, use `evalMap`. Basically don't use channels or sinks unless there really is some state that is changing over time.

Let me put together that PR and then we can experiment with this and see how it goes.

**Comment by pchiusano** [comment]
I'm going to close this issue, but we can keep talking either here or on #515.
